packages/connect/src/core/index.ts

```diff
@@ -1149,3 +1162,341 @@ const disableWebUSBTransport = async () => {
         // do nothing
     }
 };
+
+// onCallFirmwareUpdate imports
+import { randomBytes } from 'crypto';
+import {
+    getBinary,
+    uploadFirmware,
+    getLanguage,
+    calculateFirmwareHash,
+    shouldStripFwHeaders,
+    stripFwHeaders,
+} from '../api/firmware';
+import { TypedError } from '../constants/errors';
+import { getReleases } from '../data/firmwareInfo';
+import { PROTO } from '../constants';
+import { CommonParams } from '..';
+import type { Log } from '../utils/debug';
+
+export type Params = {
+    language?: string;
+    baseUrl?: string;
+    btcOnly?: boolean;
+    binary?: ArrayBuffer;
+} & CommonParams;
+
+export const onCallFirmwareUpdate = async ({
+    params,
+    deviceList,
+    postMessage,
+    log,
+}: {
+    params: Params;
+    deviceList: DeviceList;
+    postMessage: (message: CoreEventMessage) => void;
+    log: Log;
+}) => {
+    const waitForReconnectedDevice = async (
+        deviceList: DeviceList,
+        bootloader: boolean,
+    ): Promise<Device> => {
+        try {
+            await new Promise(resolve => {
+                setTimeout(() => {
+                    deviceList.enumerate().then(() => {
+                        resolve(undefined);
+                    });
+                }, 2000);
+            });
+
+            const reconnectedDevice = deviceList.getDevice(deviceList.getFirstDevicePath());
+            await reconnectedDevice.waitForFirstRun();
+            if (bootloader && !reconnectedDevice.features?.bootloader_mode) {
+                throw new Error('not in requested mode');
+            }
+            if (!bootloader && reconnectedDevice.features?.bootloader_mode) {
+                throw new Error('not in requested mode');
+            }
+
+            return reconnectedDevice;
+        } catch (err) {
+            return waitForReconnectedDevice(deviceList, bootloader);
+        }
+    };
```

Comment:
I've tried to simplify these. Is this equivalent?
```suggestion
    const waitForReconnectedDevice = async (
        deviceList: DeviceList,
        bootloader: boolean,
    ): Promise<Device> => {
        let device: Device | undefined;
        do {
            await createTimeoutPromise(2000);
            try {
                await deviceList.enumerate();
                const devicePath = deviceList.getFirstDevicePath();
                const reconnectedDevice = deviceList.getDevice(devicePath);
                await reconnectedDevice.waitForFirstRun();
                device = reconnectedDevice;
            } catch {}
        } while (!device || bootloader === !device.features?.bootloader_mode);

        return device;
    };
```